File: mobile/test/tour_functions_test.dart
import 'package:aetherride_mobile/domain/bike.dart';
import 'package:aetherride_mobile/domain/community/community_seed.dart';
import 'package:aetherride_mobile/domain/tours/tour_function_copy.dart';
import 'package:aetherride_mobile/domain/tours/tour_functions.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('reference tour exposes every function including event and club', () {
    const cats = [
      BikeCategory.gravel,
      BikeCategory.road,
      BikeCategory.etrekking,
      BikeCategory.urban,
    ];
    final events = eventsForTour(referenceTourId);
    expect(events, hasLength(1));
    expect(events.first.id, 'ev-neckar-voll');
    final clubs = clubsForTour(
      tourId: referenceTourId,
      regionSlug: 'rhein-neckar',
      categories: cats,
    );
    expect(clubs.any((c) => c.id == 'cl-rn-allround'), isTrue);
    final states = tourFunctionStates(
      tourId: referenceTourId,
      regionSlug: 'rhein-neckar',
      categories: cats,
    );
    expect(states.length, tourFunctionIds.length);
    expect(states.every((s) => s.available), isTrue);
  });

  test('every editorial event points at a catalog tour id', () {
    for (final event in communityEventSeeds) {
      expect(event.catalogTourId, isNotNull);
      expect(event.regionSlug, isNotEmpty);
    }
    expect(eventsForTour('r-heidelberg-city').first.id, 'ev-city-hd');
    expect(eventsForTour('missing'), isEmpty);
    expect(eventsForRegionSlug('rhein-neckar').length, greaterThanOrEqualTo(2));
  });

  test('sport match mirrors web catalog filters', () {
    const cats = [
      BikeCategory.gravel,
      BikeCategory.road,
      BikeCategory.etrekking,
      BikeCategory.urban,
    ];
    expect(tourMatchesSport(cats, 'gravel'), isTrue);
    expect(tourMatchesSport(cats, 'road'), isTrue);
    expect(tourMatchesSport(cats, 'urban'), isTrue);
    expect(tourMatchesSport(cats, 'hiking'), isFalse);
    expect(tourMatchesSport(cats, 'mtb'), isFalse);
  });

  test('copy keys cover every function id', () {
    for (final lang in ['de', 'en', 'fr', 'it', 'nl']) {
      final copy = tourFunctionCopy(lang);
      for (final id in tourFunctionIds) {
        expect(copy.label(id), isNot(id), reason: '$lang $id');
      }
    }
  });
}

File: src/lib/tours/tourFunctions.test.ts
/**
 * Run: npx tsx src/lib/tours/tourFunctions.test.ts
 */
import assert from "node:assert/strict";
import { getPublicTour, listPublicTours } from "../catalog/publicTours";
import { COMMUNITY_EVENTS } from "../community/seed";
import {
  REFERENCE_TOUR_ID,
  TOUR_FUNCTION_IDS,
  clubsForTour,
  eventsForRegion,
  eventsForTour,
  tourFunctionStates,
  tourHrefForEvent,
  tourMatchesSport,
} from "./tourFunctions";

function testReferenceTourHasFullKit() {
  const tour = getPublicTour(REFERENCE_TOUR_ID);
  assert.ok(tour, `missing ${REFERENCE_TOUR_ID}`);
  assert.equal(tour!.regionSlug, "rhein-neckar");
  assert.ok(tour!.loop);
  assert.ok(
    Math.abs(tour!.center[0] - 8.693) < 0.01 &&
      Math.abs(tour!.center[1] - 49.412) < 0.01,
    `reference pin ${tour!.center}`,
  );
  const events = eventsForTour(REFERENCE_TOUR_ID);
  assert.equal(events.length, 1);
  assert.equal(events[0]?.id, "ev-neckar-voll");
  const clubs = clubsForTour(tour!);
  assert.ok(clubs.some((c) => c.id === "cl-rn-allround"));
  const states = tourFunctionStates(tour!);
  assert.equal(states.length, TOUR_FUNCTION_IDS.length);
  assert.ok(states.every((s) => s.available), "reference tour must expose all functions");
}

function testEventsAttachToCatalogTours() {
  for (const event of COMMUNITY_EVENTS) {
    assert.ok(event.catalogTourId, `${event.id} missing catalogTourId`);
    assert.ok(
      getPublicTour(event.catalogTourId!),
      `${event.id} → missing tour ${event.catalogTourId}`,
    );
    assert.equal(tourHrefForEvent(event), `/tours/${event.catalogTourId}`);
  }
  assert.ok(eventsForTour("r-heidelberg-city").some((e) => e.id === "ev-city-hd"));
  assert.equal(eventsForTour("unknown-tour").length, 0);
  assert.ok(eventsForRegion("rhein-neckar").length >= 2);
}

function testSportMatch() {
  const tour = getPublicTour(REFERENCE_TOUR_ID)!;
  assert.equal(tourMatchesSport(tour, "gravel"), true);
  assert.equal(tourMatchesSport(tour, "road"), true);
  assert.equal(tourMatchesSport(tour, "urban"), true);
  assert.equal(tourMatchesSport(tour, "hiking"), false);
  assert.equal(tourMatchesSport(tour, "mtb"), false);
}

function testCatalogStillDense() {
  assert.ok(listPublicTours().length >= 90, "catalog should include the reference tour");
}

testReferenceTourHasFullKit();
testEventsAttachToCatalogTours();
testSportMatch();
testCatalogStillDense();
console.log("tourFunctions.test.ts OK");
